[DE] Add `Contribution_guide` and all subarticles (#7984)
ppy/osu-wiki: wiki/osu!_wiki/Contribution_guide/Best_practices/de.md

# Bewährte Methoden

Diese Seite beschäftigt sich mit einigen Aufgaben, die beim Beitragen zu erledigen sind. Die hier verfolgten Ansätze sind dazu gedacht, den Prozess einfacher zu machen und können auch auf andere Projekte angewandt werden, die auf GitHub oder einer ähnlichen Plattform untergebracht sind.

## Einführung

*Für mehr Informationen über git und GitHub, siehe [GitHub-Dokumentation (auf Englisch)](https://docs.github.com)*

**Git** ist ein System zur Versionsverwaltung, das dabei hilft, Änderungen an Dateien zu verwalten. Die Daten und Änderungsgeschichte des osu!-Wikis werden in einem Git-Repository gespeichert. **GitHub** ist eine Entwicklungsplattform, die eine Webschnittstelle für Git-Repositorys bietet und eine Reihe von Projektmanagement-Tools bereitstellt.

## Die Verzweigung synchronisieren

> Um Änderungen an einem Repository auf GitHub vorzunehmen, muss sich ein Mitwirkender eine isolierte Kopie des Repositorys beschaffen, welche *Verzweigung* (Fork) genannt wird. Wenn du deine Verzweigung des Repositorys `osu-wiki` erstellst, fertigst du eine Momentaufnahme des Inhalts zu diesem Augenblick an. **Synchronisiere immer deine Verzweigung** zum Beitragen, bevor du Änderungen vornimmst — das kann direkt in GitHub gemacht werden:

Reviewer: ```suggestion
Um Änderungen an einem Repository auf GitHub vorzunehmen, muss sich ein Mitwirkender eine isolierte Kopie des Repositorys beschaffen, welche *Verzweigung* (Fork) genannt wird. Wenn du deine Verzweigung des Repositorys `osu-wiki` erstellst, fertigst du eine Momentaufnahme des Inhalts zu diesem Augenblick an. Um einen sinnhaften Beitrag zu erstellen, **synchronisiere immer deine Verzweigung**, bevor du Änderungen vornimmst — das kann direkt in GitHub gemacht werden:
```

1. Gehe zu deiner Verzweigung des Repositorys `osu-wiki`.
2. Wähle den Branch `master` aus dem Dropdown-Menü aus.
3. Klicke auf `Fetch upstream` und wähle `Fetch and merge`.

![](img/update-branch.png "Den nicht mehr aktuellen Branch aktualisieren")

> Jetzt ist dein Branch auf dem neusten Stand mit dem ursprünglichen Repository.

Reviewer: ```suggestion
Jetzt ist dein Branch auf dem selben Stand wie das ursprüngliche Repository.
```

🤔

Author: Sounds good, though `dem selben` should be one word I think.

Reviewer: Fixed.

- [Synchronisiere den Branch `master`](#die-verzweigung-synchronisieren).
- Starte die Arbeit immer mit dem Erstellen eines neuen Branches, der aus dem Branch `master` hervorgeht, und mache nur dort Änderungen. Vergib einen sinnvollen Namen wie `update-staff-log`.<!-- https://www.atlassian.com/git/tutorials/comparing-workflows/forking-workflow is the explanation, but it doesn't really fit in here -->
- Nehme deine Arbeit in die Versionshistorie auf, sobald du vernünftig bemessene Änderungen gemacht hast. Es ist besser, einen Artikel als Ganzes zu übertragen, als 10 kleine Änderungen vorzunehmen.
- **Verwende kurze und sinnvolle Commit-Nachrichten**, da sie andere wissen lassen, was sie erwartet. So etwas wie `Rewrite the section about jump patterns` sagt viel mehr aus als `Update en.md`.

## Eine Änderungsanfrage erstellen

> Eine Änderungsanfrage (Pull-Request) zeigt anderen Personen, wie deine Bearbeitung die Dateien beeinflusst. Füge ein paar Informationen zu deiner Anfrage hinzu, um deine Intention zu erklären:

Reviewer: ```suggestion
Eine Änderungsanfrage (Pull-Request) zeigt anderen Personen, wie deine Änderungen die Dateien beeinflussen. Füge ein paar Informationen zu deiner Anfrage hinzu, um deine Intention zu erklären:
```

Je nach Schweregrad der Konflikte hast du zwei Möglichkeiten, diese zu beheben:

1. Wenn bei deiner Änderungsanfrage der Button `Resolve conflicts` angezeigt wird, dann klicke diesen an. Dadurch wird eine leicht andere Version des Web-Editors angezeigt.

   1. GitHub hebt Konfliktbereiche hervor. Finde eine davon.
   2. Alles von `<<<<<<<` bis `=======` sind deine Änderungen, wohingegen alles von `=======` bis `>>>>>>> master` ist, was der Branch `ppy/master` enthält.
   3. Jetzt musst du den Konflikt manuell beheben und die Zeilen mit den Markierungen `<<<<<<<`, `=======` und `>>>>>>> master` löschen.
   4. Wiederhole den Prozess für alle Konflikte.
   5. Klicke nach Abschluss auf `Mark as resolved` (das ist nur verfügbar, wenn alle Konfliktbereiche der Datei beseitigt wurden).

2. Falls der Button `Resolve conflicts` gesperrt ist, weil die Konflikte für GitHub zu kompliziert sind, hast du Pech gehabt und musst [deinen Branch aktualisieren](#die-verzweigung-synchronisieren) und deine Änderungen nochmal machen.

>    - *Anmerkung: Das stimmt nur, wenn du in der Verwendung der GitHub-Webschnittstelle eingeschränkt bist.* Es gibt immer noch Möglichkeiten, das Problem zu lösen, aber sie gehören nicht in den Rahmen dieses Artikels. Außerdem ist es vermutlich den Aufwand nicht wert, da du trotzdem die widersprüchlichen Änderungen überschreiben und rückgängig machen musst.

Reviewer: ```suggestion
   - *Anmerkung: Das stimmt nur, wenn du auf die Verwendung der GitHub-Webschnittstelle eingeschränkt bist.* Es gibt immer noch Möglichkeiten, das Problem zu lösen, aber sie gehören nicht in den Rahmen dieses Artikels. Außerdem ist es vermutlich den Aufwand nicht wert, da du trotzdem die widersprüchlichen Änderungen überschreiben und rückgängig machen musst.
```

Author: I think `beschränkt` would sound better together with `auf die` 🤔

Reviewer: Sure.
